core/src/main/java/org/infinispan/commands/triangle/BackupWriteCommand.java

```diff
@@ -0,0 +1,142 @@
+ * <p>
+ * This is a base command with the {@link CommandInvocationId}, topology and flags.
+ * <p>
+ * Since the primary->backup operations are ordered by segment, it contains the segment to be updated and its sequence number.
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public abstract class BackupWriteCommand extends BaseRpcCommand {
+
+   //common attributes of all write commands
+   private CommandInvocationId commandInvocationId;
+   private int topologyId;
+   private long flags;
+
+   //backup commands are ordered by segment. this is the sequence number of the segment.
+   private long sequence;
+   private int segmentId = -1;
+
+   private InvocationContextFactory invocationContextFactory;
+   private AsyncInterceptorChain interceptorChain;
+   private KeyPartitioner keyPartitioner;
+
+   BackupWriteCommand(ByteString cacheName) {
+      super(cacheName);
+   }
+
+   @Override
+   public final CompletableFuture<Object> invokeAsync() {
+      WriteCommand command = createWriteCommand();
+      command.setFlagsBitSet(flags);
+      command.addFlags(FlagBitSets.SKIP_LOCKING);
+      command.setValueMatcher(MATCH_ALWAYS);
+      command.setTopologyId(topologyId);
+      InvocationContext invocationContext = createContext(command);
+      return interceptorChain.invokeAsync(invocationContext, command);
+   }
+
+   @Override
+   public final boolean isReturnValueExpected() {
+      return false;
+   }
+
+   @Override
+   public final boolean canBlock() {
+      return true;
+   }
+
+   @Override
+   public void writeTo(ObjectOutput output) throws IOException {
+      CommandInvocationId.writeTo(output, commandInvocationId);
+      output.writeInt(topologyId);
+      output.writeLong(flags);
+      output.writeLong(sequence);
+   }
+
+   @Override
+   public void readFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+      commandInvocationId = CommandInvocationId.readFrom(input);
+      topologyId = input.readInt();
+      flags = input.readLong();
+      sequence = input.readLong();
+   }
+
+   public final long getSequence() {
+      return sequence;
+   }
+
+   public final void setSequence(long sequence) {
+      this.sequence = sequence;
+   }
+
+   public final CommandInvocationId getCommandInvocationId() {
+      return commandInvocationId;
+   }
+
+   public final int getTopologyId() {
+      return topologyId;
+   }
+
+   public final long getFlags() {
+      return flags;
+   }
+
+   public final int getSegmentId() {
+      if (segmentId == -1) {
+         segmentId = keyPartitioner.getSegment(anyKey());
+      }
+      return segmentId;
+   }
+
+   void setCommonAttributesFromCommand(WriteCommand command) {
+      this.commandInvocationId = command.getCommandInvocationId();
+      this.topologyId = command.getTopologyId();
+      this.flags = command.getFlagsBitSet();
+   }
+
+   abstract WriteCommand createWriteCommand();
+
+   abstract Object anyKey();
```

Comment:
This method looks weird. I wonder if we should put the segmentId to constructor (and send it over the wire), or at least define the getSegmentId() separately for each command... Take a second thought about this, please.